src/lib/common/sol-mainloop-common.c

```diff
@@ -296,6 +347,177 @@ sol_mainloop_common_idler_first(void)
     return NULL;
 }
 
+/* must be called with mainloop lock HELD */
+static inline void
+source_cleanup(void)
+{
+    struct sol_mainloop_source_common *source;
+    uint16_t i;
+
+    if (!source_pending_deletion)
+        return;
+
+    // Walk backwards so deletion doesn't impact the indices.
+    SOL_PTR_VECTOR_FOREACH_REVERSE_IDX (&source_vector, source, i) {
+        if (!source->remove_me)
+            continue;
+
+        sol_ptr_vector_del(&source_vector, i);
+        sol_mainloop_impl_unlock();
+
+        if (source->type->dispose)
+            source->type->dispose((void *)source->data);
+        free(source);
+
+        sol_mainloop_impl_lock();
+        source_pending_deletion--;
+        if (!source_pending_deletion)
+            break;
+    }
+}
+
+bool
+sol_mainloop_common_source_prepare(void)
+{
+    struct sol_mainloop_source_common *source;
+    uint16_t i;
+    bool ready = false;
+
+    sol_mainloop_impl_lock();
+    sol_ptr_vector_steal(&SOURCE_PROCESS, &SOURCE_ACUM);
+    source_processing = true;
+    sol_mainloop_impl_unlock();
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&SOURCE_PROCESS, source, i) {
+        if (!sol_mainloop_common_loop_check())
+            break;
+        if (source->remove_me)
+            continue;
+        if (source->type->prepare) {
+            source->ready = source->type->prepare((void *)source->data);
+            ready |= source->ready;
+        } else
+            source->ready = false;
+    }
+
+    sol_mainloop_impl_lock();
+    sol_ptr_vector_update(&SOURCE_ACUM, &SOURCE_PROCESS);
+    source_cleanup();
+    source_processing = false;
+    sol_mainloop_impl_unlock();
+
+    return ready;
+}
+
+/* must be called with mainloop lock HELD */
+static bool
+sol_mainloop_common_source_get_next_timeout_locked(struct timespec *timeout)
+{
+    bool found = false;
+    struct sol_mainloop_source_common *source;
+    uint16_t i;
+
+    source_processing = true;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&SOURCE_PROCESS, source, i) {
+        struct timespec cur;
+        bool r;
+
+        if (source->remove_me)
+            continue;
+
+        if (!source->type->get_next_timeout)
+            continue;
+
+        r = source->type->get_next_timeout((void *)source->data, &cur);
+        if (r) {
+            if (!found || sol_util_timespec_compare(&cur, timeout) < 0) {
+                if (cur.tv_sec < 0) {
+                    cur.tv_sec = 0;
+                    cur.tv_nsec = 0;
+                }
+                *timeout = cur;
+            }
+            found = true;
+        }
+    }
+
+    source_processing = false;
+
+    return found;
+}
+
+bool
+sol_mainloop_common_source_get_next_timeout(struct timespec *timeout)
+{
+    bool r;
+
+    sol_mainloop_impl_lock();
```

Comment:
could be, but this one is really an exception in usage since we don't want to keep it locked for long time, see the other functions.

This one is not even used, I just added it for completeness, only the _locked variant is used at the moment.